File: ecs-hosted/ts/application/logs/awsLogs.ts
import * as pulumi from "@pulumi/pulumi";
import * as cloudwatch from "@pulumi/aws/cloudwatch";
import { LogDriver, LogType } from "./types";

export interface AwsLogsArgs {
    retentionInDays?: number,
    name: string,
    region: string
}

const namespace = "pulumi:cloudwatch-aws-logs";

/*
ComponentResource built to create AWS Cloudwatch Log Group
*/
export class AwsLogs extends pulumi.ComponentResource implements LogDriver {
    public logType: LogType = LogType.awslogs;
    public readonly logGroup: cloudwatch.LogGroup;

    // Allow concrete implementation to register output values that can be resolved (applied) through the interface
    public outputs: pulumi.Output<string[]>;
    private readonly args: AwsLogsArgs;

    constructor(name: string, args: AwsLogsArgs, opts?: pulumi.ComponentResourceOptions) {
        super(namespace, name, opts);

        this.args = args;

        this.logGroup = new cloudwatch.LogGroup(`${name}-cw-lg`, {
            name: `${args.name}-${name}-logs`,
            retentionInDays: args.retentionInDays ? args.retentionInDays! : 7
        }, opts);

        this.outputs = pulumi.output([this.logGroup.id]);
    }

    // return specific "configuration" for this types logs
    getConfiguration(ids: string[]): any {
        return {
            "awslogs-region": this.args.region,
            "awslogs-group": ids[0],
            "awslogs-stream-prefix": "awslogs-pulumi",
        };
    }
}

File: ecs-hosted/go/application/log/awsLogs.go
package log

import (
	"fmt"
	"github.com/pulumi/pulumi-aws/sdk/v7/go/aws/cloudwatch"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func NewAwsLogs(ctx *pulumi.Context, name string, args *AwsArgs, opts ...pulumi.ResourceOption) (*AwsLogs, error) {
	lg, err := cloudwatch.NewLogGroup(ctx, fmt.Sprintf("awslogs-%s", name), &cloudwatch.LogGroupArgs{
		NamePrefix:      pulumi.String(fmt.Sprintf("cloudwatch-%s-logs", name)),
		RetentionInDays: pulumi.Int(args.RetentionDays),
	}, opts...)

	if err != nil {
		return nil, err
	}

	resource := AwsLogs{
		LogGroup: lg,
		Region:   args.Region,
		Outputs: map[string]any{
			"region":     args.Region,
			"logGroupId": lg.ID(),
		},
	}

	return &resource, nil
}

func (l AwsLogs) GetConfiguration() map[string]any {

	return map[string]any{
		"logDriver": "awslogs",
		"options": map[string]any{
			"awslogs-region":        l.Outputs["region"],
			"awslogs-group":         l.Outputs["logGroupId"],
			"awslogs-stream-prefix": "pulumi-api",
		},
	}
}

type AwsArgs struct {
	Region        string
	Name          string
	RetentionDays int
}

type AwsLogs struct {
	Region   string
	LogGroup *cloudwatch.LogGroup
	Outputs  map[string]any
}
